promql/engine.go

```diff
@@ -706,117 +654,354 @@ func (ev *evaluator) eval(expr Expr) Value {
+			for i := range exprs {
+				vectors[i] = vectors[i][:0]
+				for si, series := range matrixes[i] {
+					for pos, point := range series.Points {
+						if point.T == ts {
+							vectors[i] = append(vectors[i], Sample{Metric: series.Metric, Point: point})
+						}
+						if point.T >= ts {
+							// Move input vectors forward so we don't have to re-scan the same
+							// past points at the next step.
+							matrixes[i][si].Points = series.Points[pos:]
+							break
+						}
+					}
+				}
+				args[i] = vectors[i]
+			}
+			// Make the function call.
+			enh.ts = ts
+			result := f(args, enh)
+			enh.out = result[:0] // Reuse result vector.
+			// If this could be an instant query, shortcut so as not to change sort order.
+			if ev.endTimestamp == ev.timestamp {
+				mat := make(Matrix, len(result))
+				for i, s := range result {
+					s.Point.T = ts
+					mat[i] = Series{Metric: s.Metric, Points: []Point{s.Point}}
+				}
+				return mat
+			}
+			// Add samples in output vector to output series.
+			for _, sample := range result {
+				h := sample.Metric.Hash()
+				ss, ok := seriess[h]
+				if !ok {
+					ss = Series{
+						Metric: sample.Metric,
+						Points: getPointSlice(numSteps),
+					}
+					seriess[h] = ss
+				}
+				sample.Point.T = ts
+				ss.Points = append(ss.Points, sample.Point)
+				seriess[h] = ss
+			}
+		}
+		// Reuse the original point slices.
+		for _, m := range origMatrixes {
+			for _, s := range m {
+				putPointSlice(s.Points)
+			}
+		}
+		// Assemble the output matrix.
+		mat := Matrix{}
+		for _, ss := range seriess {
+			mat = append(mat, ss)
+		}
+		return mat
+	}
 
 	switch e := expr.(type) {
 	case *AggregateExpr:
-		Vector := ev.evalVector(e.Expr)
-		return ev.aggregation(e.Op, e.Grouping, e.Without, e.Param, Vector)
+		if s, ok := e.Param.(*StringLiteral); ok {
+			return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+				return ev.aggregation(e.Op, e.Grouping, e.Without, s.Val, v[0].(Vector), enh)
+			}, e.Expr)
+		}
+		return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+			var param float64
+			if e.Param != nil {
+				param = v[0].(Vector)[0].V
+			}
+			return ev.aggregation(e.Op, e.Grouping, e.Without, param, v[1].(Vector), enh)
+		}, e.Param, e.Expr)
 
-	case *BinaryExpr:
-		lhs := ev.evalOneOf(e.LHS, ValueTypeScalar, ValueTypeVector)
-		rhs := ev.evalOneOf(e.RHS, ValueTypeScalar, ValueTypeVector)
+	case *Call:
+		if e.Func.Name == "timestamp" {
+			// Matrix evaluation always returns the evaluation time,
+			// so this function needs special handling when given
+			// a vector selector.
+			vs, ok := e.Args[0].(*VectorSelector)
+			if ok {
+				return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+					return e.Func.Call([]Value{ev.vectorSelector(vs, enh.ts)}, e.Args, enh)
+				})
+			}
+		}
+		// Check if the function has a matrix argument.
+		for matrixArgIndex, a := range e.Args {
+			sel, ok := a.(*MatrixSelector)
+			if ok {
```

Comment:
This seems to go against Go style (return early to reduce nesting). The `else` case is only a few lines long, so it may make sense to handle that first. Then you won't need the 2 extra levels of nesting for the big block of code below.

Reply:
It's a bit trickier than that due to the for. I've reworked it.